sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/StorageSeekableByteChannel.java

```diff
@@ -0,0 +1,308 @@
+        buffer.clear();
+        int read = readBehavior.read(buffer, newBufferAbsolutePosition);
+        buffer.rewind();
+        buffer.limit(Math.max(read, 0));
+        bufferAbsolutePosition = Math.min(newBufferAbsolutePosition, readBehavior.getResourceLength());
+        return read;
+    }
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+        assertOpen();
+        assertCanWrite();
+
+        int write = Math.min(src.remaining(), buffer.remaining());
+        if (write > 0) {
+            ByteBuffer temp = src.duplicate();
+            temp.limit(temp.position() + write);
+            buffer.put(temp);
+            src.position(src.position() + write);
+        }
+
+        if (buffer.remaining() == 0) {
+            try {
+                flushWriteBuffer();
+            } catch (RuntimeException e) {
+                // undo write on failed flush by rewinding buffer position the amount it was incremented
+                buffer.position(buffer.position() - write);
+                throw LOGGER.logExceptionAsError(e);
+            }
+        }
+
+        absolutePosition += write;
+        return write;
+    }
+
+    private void flushWriteBuffer() {
+        if (buffer.position() == 0) {
+            return;
+        }
+
+        int startingPosition = buffer.position();
+        buffer.limit(buffer.position());
+        buffer.rewind();
+        try {
+            writeBehavior.write(buffer, bufferAbsolutePosition);
+        } catch (RuntimeException e) {
+            // restore buffer state if write fails
+            buffer.limit(buffer.capacity());
+            buffer.position(startingPosition);
+            throw LOGGER.logExceptionAsError(e);
+        }
+        bufferAbsolutePosition += buffer.limit();
+        buffer.clear();
+    }
+
+    @Override
+    public long position() throws IOException {
+        assertOpen();
+        return absolutePosition;
+    }
+
+    @Override
+    public SeekableByteChannel position(long newPosition) throws IOException {
+        assertOpen();
+
+        if (readBehavior != null) {
+            readModeSeek(newPosition);
+        } else {
+            writeModeSeek(newPosition);
+        }
+
+        return this;
+    }
+
+    private void readModeSeek(long newPosition) {
+        // seek exited the bounds of the internal buffer, invalidate it
+        if (newPosition < bufferAbsolutePosition || newPosition > bufferAbsolutePosition + buffer.limit()) {
+            buffer.clear();
+            buffer.limit(0);
+        // seek is within the internal buffer, just adjust buffer position
+        } else {
+            buffer.position((int) (newPosition - bufferAbsolutePosition));
+        }
+        absolutePosition = newPosition;
+    }
+
+    private void writeModeSeek(long newPosition) {
+        if (!writeBehavior.canSeek(newPosition)) {
+            throw LOGGER.logExceptionAsError(new IllegalArgumentException(
```

Comment:
You're suggesting `assertCanSeek()` on the interface?

Reply:
Yeah, something like that, since the boolean is only used to determine whether this exception should be thrown it may be better to make it `void` and throw a more detailed exception than what's being thrown here.